efficientdet/anchors.py

```diff
@@ -139,7 +140,80 @@ def nms(dets, thresh):
 
     inds = np.where(overlap <= thresh)[0]
     order = order[inds + 1]
-  return keep
+
+  return dets[keep]
+
+
+def softnms(dets, nms_configs):
+  """Soft non-maximum suppression.
+
+  [1] Soft-NMS -- Improving Object Detection With One Line of Code.
+    https://arxiv.org/abs/1704.04503
+
+  Args:
+    dets: detection with shape (num, 5) and format [x1, y1, x2, y2, score].
+    nms_configs: a dict config that may contain the following members
+      * method: one of {`linear`, `gaussian` or 'greedy'}. Use `greedy` if None.
+      * sigma: Gaussian parameter, only for method 'gaussian'.
+      * iou_thresh (float): IOU threshold, only for `linear` or 'greedy'.
+      * score_thresh (float): Box score threshold for final boxes.
+
+  Returns:
+    numpy.array: Retained boxes.
+  """
+  method = nms_configs.get('method', 'greedy')
+  sigma = nms_configs.get('sigma', 0.5)
+  iou_thresh = nms_configs.get('iou_thresh', 0.5)
+  score_thresh = nms_configs.get('score_thresh', 0.001)
+
+  if method not in ('linear', 'gaussian', 'greedy'):
+    raise ValueError(
+        'NMS method must be linear/gaussian/greedy, got: {}'.format(method))
+
+  if method == 'greedy' or not method:
+    # the default nms has the same output as greedy method, but runs faster.
+    return nms(dets, iou_thresh)
+
+  x1 = dets[:, 0]
+  y1 = dets[:, 1]
+  x2 = dets[:, 2]
+  y2 = dets[:, 3]
+
+  areas = (x2 - x1 + 1) * (y2 - y1 + 1)
+  # expand dets with areas, and the second dimension is
+  # x1, y1, x2, y2, score, area
+  dets = np.concatenate((dets, areas[:, None]), axis=1)
+
+  retained_box = []
+  while dets.size > 0:
+    max_idx = np.argmax(dets[:, 4], axis=0)
+    dets[[0, max_idx], :] = dets[[max_idx, 0], :]
+    retained_box.append(dets[0, :-1])
+
+    xx1 = np.maximum(dets[0, 0], dets[1:, 0])
+    yy1 = np.maximum(dets[0, 1], dets[1:, 1])
+    xx2 = np.minimum(dets[0, 2], dets[1:, 2])
+    yy2 = np.minimum(dets[0, 3], dets[1:, 3])
+
+    w = np.maximum(xx2 - xx1 + 1, 0.0)
+    h = np.maximum(yy2 - yy1 + 1, 0.0)
+    inter = w * h
+    iou = inter / (dets[0, 5] + dets[1:, 5] - inter)
+
+    if method == 'linear':
+      weight = np.ones_like(iou)
+      weight[iou > iou_thresh] -= iou[iou > iou_thresh]
+    elif method == 'gaussian':
+      weight = np.exp(-(iou * iou) / sigma)
+    else:  # traditional nms
+      weight = np.ones_like(iou)
```

Comment:
Same as nms?

Reply:
yes, the output is the same, but the other nms function runs slightly faster than this one (due to the implementation).